### PR TITLE
Make it easier to add commands out-of-tree

### PR DESCRIPTION
**Pull request by martinvonz**

### CODE REVIEW

**Review by martinvonz — COMMENTED** [review]
Actually, I should add some tests for the example command. We know that it's not hard to call it from tests.

**Comment by martinvonz** [comment]
> Actually, I should add some tests for the example command. We know that it's not hard to call it from tests.

Since I was wrong about how `cargo test` also builds `examples/` (it doesn't), it's not easy to test the example command after all :( So this is ready for review.

File: src/main.rs [hunk 1]
```diff
@@ -12,44 +12,24 @@
...
-                }
-                Err(CommandError::InternalError(message)) => {
-                    ui.write_error(&format!("Internal error: {}\n", message))
-                        .unwrap();
-                    std::process::exit(255);
-                }
-            }
+    let mut ui = create_ui();
+    match run(&mut ui) {
+        Ok(_) => {
```
Comment (arxanas) [inline src/main.rs:28]:
```suggestion
        Ok(()) => {
```

Comment (martinvonz) [reply src/main.rs]: Fixed

**Review by yuja — APPROVED** [review]

**Comment by yuja** [comment]
It seems a bit scary that process::exit() paths are spreading around.
Maybe we can add a high level helper that takes `run_command()`-like closure, but I don't know that is a good abstraction level.

**Comment by martinvonz** [comment]
> It seems a bit scary that process::exit() paths are spreading around. Maybe we can add a high level helper that takes `run_command()`-like closure, but I don't know that is a good abstraction level.

Yes, true. An earlier version of this PR had `report_command_error()` return a `std::process::ExitCode` but that failed the Nix build because Nixpkgs is on 1.60 and `std::process::ExitCode` came in 1.61. I've added a TODO about switching to `std::process::ExitCode` once our MSRV is 1.61. Sounds good?

**Comment by martinvonz** [comment]
> > It seems a bit scary that process::exit() paths are spreading around. Maybe we can add a high level helper that takes `run_command()`-like closure, but I don't know that is a good abstraction level.
> 
> Yes, true. An earlier version of this PR had `report_command_error()` return a `std::process::ExitCode` but that failed the Nix build because Nixpkgs is on 1.60 and `std::process::ExitCode` came in 1.61. I've added a TODO about switching to `std::process::ExitCode` once our MSRV is 1.61. Sounds good?

Oh, but there's `create_ui()` that also calls `process::exit()`. Hmm, let me see what I can do.

**Comment by yuja** [comment]
> but there's `create_ui()` that also calls `process::exit()`.

Yes, and `resolve_aliases()` would also do via `error().exit()`.

**Comment by martinvonz** [comment]
> > but there's `create_ui()` that also calls `process::exit()`.
> 
> Yes, and `resolve_aliases()` would also do via `error().exit()`.

True, and even inside `clap::Command::get_matches_from()`. We need to switch to `try_get_matches_from()` we want to handle system exit ourselves.

For now, I made `report_command_error()` return an int32. I'll look into cleaning up the other exit calls after this series.

**Pull request merged**